README.md
# Row filter plugin for Embulk

[![Build Status](https://secure.travis-ci.org/sonots/embulk-filter-row.png?branch=master)](http://travis-ci.org/sonots/embulk-filter-row)

A filter plugin for Embulk to filter out rows

## Configuration

Requirement: version >= 0.3.0

* **where**: Select only rows which match with conditions written in SQL-like syntax. See [SQL-like Syntax](#sql-like-syntax)

## Example

```yaml
filters:
  - type: row
    where: column1 = 'str'
```

```yaml
filters:
  - type: row
    where: |-
      (
        string_column START_WITH 'str' AND
        number_column > 1.0
      )
      OR
      (
        time_column = TIMESTAMP '2016-01-01 +0900' AND
        "true_column" = true
      )
```

See [SQL-like Syntax](#sql-like-syntax) for more details

# SQL-like Syntax

This syntax must be similar with a standard SQL syntax.

```sql
where: column1 = 'str'
```

```sql
where: |-
  (
    string_column START_WITH 'str' AND
    number_column > 1.0
  
  )
  OR
  (
    time_column = TIMESTAMP '2016-01-01 +0900' AND
    "true_column" = true AND
    string_column REGEXP '^reg'
  )
```

## Literals

### Boolean Literal

`true` or `TRUE` or `false` or `FALSE` are considered as a boolean literal

### Number Literal

Characters matching with a regular expression `-?[0-9]+(\.[0-9]+)?` is considered as a number literal

### String Literal

Characters surrounded by `'` such as `'foo'` is considered as a string literal

### Timestamp Literal

NOTE: It became possible to omit `TIMESTAMP` keyword on comparing with `timestamp` identifier (column) from version >= 0.3.3.

`TIMESTAMP ( NumberLiteral | StringLiteral )` such as `TIMESTAMP 1470433087.747123` or `TIMESTAMP '2016-08-06 06:38:07.747123 +0900'` is considered as a timestamp literal

Number is a epoch time since 1970-01-01 UTC with nano time resolution.

String is a timestamp string which matches with one of following format:

* `%Y-%m-%d %H:%M:%S.%N %z`
* `%Y-%m-%d %H:%M:%S.%N`
* `%Y-%m-%d %H:%M:%S %z`
* `%Y-%m-%d %H:%M:%S`
* `%Y-%m-%d %z`
* `%Y-%m-%d`

The time zone for formats without `%z` is UTC, and the time resolution is micro second (caused by limitation of Embulk TimestampParser).

### Json Literal

Not supported yet

### Identifier Literal

Characters matching with a regular expression `[a-zA-Z_][a-zA-z0-9_]*` such as `foobar`, and characters surrounded by `"` such as `"foo-bar"`, `"foo.bar"`, and `"foo\"bar"` are considred as an identifier literal, that is, embulk's column name.

## Operators

### Boolean Operator

* `=`
* `!=`

### Number Operator (Long and Double)

* `=`
* `!=`
* `>`
* `>=`
* `<=`
* `<`

### String Operator

* `=`
* `!=`
* `START_WITH`
* `END_WITH`
* `INCLUDE`
* `REGEXP`

### Timestamp Operator

* `=`
* `!=`
* `>`
* `>=`
* `<=`
* `<`

### Json Operator

Not supported yet

### unary operator

* "xxx IS NULL"
* "xxx IS NOT NULL"
* "NOT xxx"

## Old Configuration

Versions >= 0.3.0 has `where` option to supports SQL-like syntax. I recommend to use it.

Following options are **deprecated**, and **will be removed someday**.

* **condition**: AND or OR (string, default: AND).
* **conditions**: select only rows which matches with conditions.
  * **column**: column name (string, required)
  * **operator** operator (string, optional, default: ==)
    * boolean operator
      * `==`
      * `!=`
    * numeric operator (long, double, Timestamp)
      * `==`
      * `!=`
      * `>`
      * `>=`
      * `<=`
      * `<`
    * string operator
      * `==`
      * `!=`
      * `start_with` (or `startsWith`)
      * `end_with` (or `endsWith`)
      * `include` (or `contains`)
    * unary operator
      * `IS NULL`
      * `IS NOT NULL`
  * **argument**: argument for the operation (string, required for non-unary operators)
  * **not**: not (boolean, optional, default: false)
  * **format**: special option for timestamp column, specify the format of timestamp argument, parsed argument is compared with the column value as Timestamp object (string, default is `%Y-%m-%d %H:%M:%S.%N %z`)
  * **timezone**: special option for timestamp column, specify the timezone of timestamp argument (string, default is `UTC`)

NOTE: column type is automatically retrieved from input data (inputSchema)

## Example (AND)

**Deprecated**

```yaml
filters:
  - type: row
    condition: AND
    conditions:
      - {column: foo,  operator: "IS NOT NULL"}
      - {column: id,   operator: ">=", argument: 10}
      - {column: id,   operator: "<",  argument: 20}
      - {column: name, opeartor: "include", argument: foo, not: true}
      - {column: time, operator: "==", argument: "2015-07-13", format: "%Y-%m-%d"}
```

## Example (OR)

**Deprecated**

```yaml
filters:
  - type: row
    condition: OR
    conditions:
      - {column: a, operator: "IS NOT NULL"}
      - {column: b, operator: "IS NOT NULL"}
```

## Example (AND of OR)

**Deprecated**

You can express a condition such as `(A OR B) AND (C OR D)` by combining multiple filters like

```yaml
filters:
  - type: row
    condition: OR
    conditions:
      - {column: a, operator: "IS NOT NULL"}
      - {column: b, operator: "IS NOT NULL"}
  - type: row
    condition: OR
    conditions:
      - {column: c, operator: "IS NOT NULL"}
      - {column: d, operator: "IS NOT NULL"}
```

## Comparisions

* [embulk-filter-calcite](https://github.com/muga/embulk-filter-calcite)
  * embulk-filter-calcite is a pretty nice plugin which enables us to write SQL query to filter embulk records, not only `WHERE` but also `SELECT`.
  * Based on [my benchmark (Japanese)](http://qiita.com/sonots/items/a70482d29862de87624d), embulk-filter-row was faster than embulk-filter-calcite.
  * Choose which to use as your demand.

## ToDo

* Support filtering by values of `type: json` with JSONPath
* Support IN operator

## ChangeLog

[CHANGELOG.md](./CHANGELOG.md)

## Development

Run example:

```
$ ./gradlew classpath
$ embulk preview -I lib example/example.yml
```

Run test:

```
$ ./gradlew test
```

Run checkstyle:

```
$ ./gradlew check
```

Release gem:

```
$ ./gradlew gemPush
```

## Development of SQL-like Syntax

Read the article [Supported SQL-like Syntax with embulk-filter-row using BYACC/J and JFlex](http://blog.livedoor.jp/sonots/archives/48172830.html).

To download BYACC/J and JFlex and run them, you can use:

```
$ script/byaccj.sh
```

or

```
$ ./gradlew byaccj # this runs script/byaccj.sh internally
```

This generates `src/main/java/org/embulk/filter/row/where/{Parser,ParserVal,Yylex}.java`.

The `byaccj` task of gradle is ran before `compileJava` task (which means to be ran before `classpath` or `test` task also) automatically.


CHANGELOG.md
# 0.6.1 (2022-10-24)

Fixes:

* Avoid NullPointerException (thanks to Tetsuro Sano)

# 0.6.0 (2021-06-10)

Changes:

* Get ready for v0.11 and v1.0 (thanks to hiroyuki-sato)

# 0.5.0 (2017-08-23)

Changes

* Follow new TimestampParser API of embulk >= 0.8.29.
  * Note that embulk-filter-row >= 0.5.0 requires embulk >= 0.8.29.

# 0.4.0 (2017-05-18)

Enhancements

* Use joni library for REGEXP operator to improve performance
  * Note that This may introduce trival incompatibility changes

# 0.3.3 (2016-08-09)

Enhancements

* Support REGEXP operator
* Allow to omit TIMESTAMP keyword

# 0.3.2 (2016-08-08)

Changes

* `conditions` option is now deprecated. Show deprecation warning messages.

Fxies:

* Fix NullPointerException for `where` option

# 0.3.1 (2016-08-08)

Changes:

* Change identifier syntax from `[a-zA-Z$][a-zA-z0-9\.\-_]*` to `[a-zA-Z_][a-zA-z0-9_]*` to allow starting _, disallow staring $, disallow -, disallow . (dot).

# 0.3.0 (2016-08-06)

Enhancements:

* Support SQL-like Syntax

# 0.2.2 (2016-08-05)

Enhancements:

* Support `type: json` (not supporting filtering with json yet, though)

# 0.2.1 (2016-06-27)

Enhancements:

* Add a workaround to treat `0.x` argument as a double which is treated as a String by sanekyaml (thanks to @toyama0919)

# 0.2.0 (2015-12-05)

Enhancements:

* Support OR condition

# 0.1.4

Fixes:

* raise SchemaConfigException if column name is not found

# 0.1.3

Fixes:

* Remove debug print

# 0.1.2

Enhancements:

* Support Java 1.7

# 0.1.1

Enhancements:

* Support `startsWith`, `endsWith`, `contains` string operators

# 0.1.0

first version


yacc/Yylex.java
/* The following code was generated by JFlex 1.6.1 */

/**
 * This class is a scanner generated by 
 * <a href="http://www.jflex.de/">JFlex</a> 1.6.1
 * from the specification file <tt>where.flex</tt>
 */
class Yylex {

  /** This character denotes the end of file */
  public static final int YYEOF = -1;

  /** initial size of the lookahead buffer */
  private static final int ZZ_BUFFERSIZE = 16384;

  /** lexical states */
  public static final int YYINITIAL = 0;
  public static final int STRING = 2;
  public static final int IDENTIFIER = 4;

  /**
   * ZZ_LEXSTATE[l] is the state in the DFA for the lexical state l
   * ZZ_LEXSTATE[l+1] is the state in the DFA for the lexical state l
   *                  at the beginning of a line
   * l is of the form l = 2*k, k a non negative integer
   */
  private static final int ZZ_LEXSTATE[] = { 
     0,  0,  1,  1,  2, 2
  };

  /** 
   * Translates characters to character classes
   */
  private static final String ZZ_CMAP_PACKED = 
    "\11\0\1\4\1\5\1\45\1\45\1\5\22\0\1\4\1\20\1\3"+
    "\4\0\1\6\1\7\1\7\4\0\1\2\1\0\12\1\2\0\1\16"+
    "\1\15\1\17\2\0\1\10\1\0\1\30\1\12\1\27\1\43\1\0"+
    "\1\26\1\25\2\0\1\31\1\0\1\11\1\13\2\0\1\14\1\21"+
    "\1\22\1\32\1\0\1\24\4\0\1\44\2\0\1\23\1\0\1\40"+
    "\1\46\2\0\1\36\1\37\5\0\1\41\1\0\1\47\3\0\1\34"+
    "\1\42\1\33\1\35\17\0\1\45\u1fa2\0\1\45\1\45\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\uffff\0\udfe6\0";

  /** 
   * Translates characters to character classes
   */
  private static final char [] ZZ_CMAP = zzUnpackCMap(ZZ_CMAP_PACKED);

  /** 
   * Translates DFA states to action switch labels.
   */
  private static final int [] ZZ_ACTION = zzUnpackAction();

  private static final String ZZ_ACTION_PACKED_0 =
    "\3\0\1\1\1\2\1\3\1\4\1\5\1\6\1\7"+
    "\3\1\1\10\1\11\1\12\10\1\1\13\1\14\1\5"+
    "\1\13\1\15\1\5\4\1\1\16\1\17\1\20\1\21"+
    "\3\1\1\22\4\1\1\23\1\24\1\25\1\26\1\27"+
    "\1\30\1\31\1\32\1\33\1\2\1\34\1\35\10\1"+
    "\1\36\1\1\1\37\7\1\1\40\4\1\1\41\2\1"+
    "\1\42\1\1\1\43";

  private static int [] zzUnpackAction() {
    int [] result = new int[87];
    int offset = 0;
    offset = zzUnpackAction(ZZ_ACTION_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackAction(String packed, int offset, int [] result) {
    int i = 0;       /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int count = packed.charAt(i++);
      int value = packed.charAt(i++);
      do result[j++] = value; while (--count > 0);
    }
    return j;
  }


  /** 
   * Translates a state to a row index in the transition table
   */
  private static final int [] ZZ_ROWMAP = zzUnpackRowMap();

  private static final String ZZ_ROWMAP_PACKED_0 =
    "\0\0\0\50\0\120\0\170\0\240\0\310\0\360\0\310"+
    "\0\310\0\170\0\u0118\0\u0140\0\u0168\0\170\0\u0190\0\u01b8"+
    "\0\u01e0\0\u0208\0\u0230\0\u0258\0\u0280\0\u02a8\0\u02d0\0\u02f8"+
    "\0\u0320\0\310\0\u0348\0\u0370\0\310\0\u0398\0\u03c0\0\u03e8"+
    "\0\u0410\0\u0438\0\170\0\170\0\170\0\170\0\u0460\0\u0488"+
    "\0\u04b0\0\170\0\u04d8\0\u0500\0\u0528\0\u0550\0\310\0\310"+
    "\0\310\0\310\0\310\0\310\0\310\0\310\0\310\0\u03c0"+
    "\0\170\0\170\0\u0578\0\u05a0\0\u05c8\0\u05f0\0\u0618\0\u0640"+
    "\0\u0668\0\u0690\0\170\0\u06b8\0\170\0\u06e0\0\u0708\0\u0730"+
    "\0\u0758\0\u0780\0\u07a8\0\u07d0\0\170\0\u07f8\0\u0820\0\u0848"+
    "\0\u0870\0\170\0\u0898\0\u08c0\0\170\0\u08e8\0\170";

  private static int [] zzUnpackRowMap() {
    int [] result = new int[87];
    int offset = 0;
    offset = zzUnpackRowMap(ZZ_ROWMAP_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackRowMap(String packed, int offset, int [] result) {
    int i = 0;  /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int high = packed.charAt(i++) << 16;
      result[j++] = high | packed.charAt(i++);
    }
    return j;
  }

  /** 
   * The transition table of the DFA
   */
  private static final int [] ZZ_TRANS = zzUnpackTrans();

  private static final String ZZ_TRANS_PACKED_0 =
    "\1\4\1\5\1\4\1\6\1\7\1\10\1\11\1\12"+
    "\1\13\1\14\1\4\1\15\1\4\1\16\1\17\1\20"+
    "\1\21\1\22\1\23\2\4\1\24\1\4\1\25\3\4"+
    "\1\26\3\4\1\27\3\4\1\30\1\10\3\4\5\31"+
    "\1\10\1\32\35\31\1\33\3\31\3\34\1\35\1\34"+
    "\1\10\36\34\1\36\3\34\3\4\4\0\35\4\1\0"+
    "\4\4\1\5\1\37\4\0\35\4\1\0\3\4\54\0"+
    "\1\7\43\0\3\4\4\0\2\4\1\40\32\4\1\0"+
    "\6\4\4\0\4\4\1\41\16\4\1\42\11\4\1\0"+
    "\6\4\4\0\5\4\1\43\27\4\1\0\6\4\4\0"+
    "\6\4\1\44\1\4\1\45\24\4\1\0\6\4\4\0"+
    "\6\4\1\46\26\4\1\0\6\4\4\0\6\4\1\45"+
    "\26\4\1\0\6\4\4\0\13\4\1\47\21\4\1\0"+
    "\6\4\4\0\5\4\1\50\27\4\1\0\6\4\4\0"+
    "\2\4\1\51\7\4\1\52\22\4\1\0\6\4\4\0"+
    "\2\4\1\53\32\4\1\0\6\4\4\0\25\4\1\54"+
    "\7\4\1\0\6\4\4\0\31\4\1\55\3\4\1\0"+
    "\6\4\4\0\1\4\1\56\33\4\1\0\3\4\5\31"+
    "\2\0\35\31\1\0\3\31\3\57\1\60\1\57\1\0"+
    "\1\61\24\57\1\62\1\63\2\57\1\64\4\57\1\65"+
    "\1\0\1\66\1\67\3\34\1\0\1\34\1\0\36\34"+
    "\1\0\3\34\3\57\1\60\1\57\1\0\1\61\35\57"+
    "\1\65\1\0\2\57\1\4\1\70\1\4\4\0\35\4"+
    "\1\0\6\4\4\0\3\4\1\71\31\4\1\0\6\4"+
    "\4\0\13\4\1\72\21\4\1\0\6\4\4\0\22\4"+
    "\1\73\12\4\1\0\6\4\4\0\1\4\1\74\33\4"+
    "\1\0\6\4\4\0\23\4\1\75\11\4\1\0\6\4"+
    "\4\0\21\4\1\76\13\4\1\0\6\4\4\0\3\4"+
    "\1\77\31\4\1\0\6\4\4\0\26\4\1\100\6\4"+
    "\1\0\6\4\4\0\32\4\1\101\2\4\1\0\6\4"+
    "\4\0\22\4\1\102\12\4\1\0\6\4\4\0\22\4"+
    "\1\103\12\4\1\0\6\4\4\0\5\4\1\104\27\4"+
    "\1\0\6\4\4\0\20\4\1\105\14\4\1\0\6\4"+
    "\4\0\22\4\1\106\12\4\1\0\6\4\4\0\14\4"+
    "\1\107\20\4\1\0\6\4\4\0\27\4\1\105\5\4"+
    "\1\0\6\4\4\0\33\4\1\110\1\4\1\0\6\4"+
    "\4\0\12\4\1\111\22\4\1\0\6\4\4\0\13\4"+
    "\1\112\21\4\1\0\6\4\4\0\23\4\1\113\11\4"+
    "\1\0\6\4\4\0\15\4\1\114\17\4\1\0\6\4"+
    "\4\0\27\4\1\115\5\4\1\0\6\4\4\0\20\4"+
    "\1\115\14\4\1\0\6\4\4\0\14\4\1\116\20\4"+
    "\1\0\6\4\4\0\3\4\1\117\31\4\1\0\6\4"+
    "\4\0\16\4\1\120\16\4\1\0\6\4\4\0\15\4"+
    "\1\121\17\4\1\0\6\4\4\0\20\4\1\122\14\4"+
    "\1\0\6\4\4\0\13\4\1\123\21\4\1\0\6\4"+
    "\4\0\16\4\1\124\16\4\1\0\6\4\4\0\17\4"+
    "\1\125\15\4\1\0\6\4\4\0\13\4\1\126\21\4"+
    "\1\0\6\4\4\0\17\4\1\127\15\4\1\0\3\4";

  private static int [] zzUnpackTrans() {
    int [] result = new int[2320];
    int offset = 0;
    offset = zzUnpackTrans(ZZ_TRANS_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackTrans(String packed, int offset, int [] result) {
    int i = 0;       /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int count = packed.charAt(i++);
      int value = packed.charAt(i++);
      value--;
      do result[j++] = value; while (--count > 0);
    }
    return j;
  }


  /* error codes */
  private static final int ZZ_UNKNOWN_ERROR = 0;
  private static final int ZZ_NO_MATCH = 1;
  private static final int ZZ_PUSHBACK_2BIG = 2;

  /* error messages for the codes above */
  private static final String ZZ_ERROR_MSG[] = {
    "Unknown internal scanner error",
    "Error: could not match input",
    "Error: pushback value was too large"
  };

  /**
   * ZZ_ATTRIBUTE[aState] contains the attributes of state <code>aState</code>
   */
  private static final int [] ZZ_ATTRIBUTE = zzUnpackAttribute();

  private static final String ZZ_ATTRIBUTE_PACKED_0 =
    "\3\0\2\1\1\11\1\1\2\11\20\1\1\11\2\1"+
    "\1\11\21\1\11\11\40\1";

  private static int [] zzUnpackAttribute() {
    int [] result = new int[87];
    int offset = 0;
    offset = zzUnpackAttribute(ZZ_ATTRIBUTE_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackAttribute(String packed, int offset, int [] result) {
    int i = 0;       /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int count = packed.charAt(i++);
      int value = packed.charAt(i++);
      do result[j++] = value; while (--count > 0);
    }
    return j;
  }

  /** the input device */
  private java.io.Reader zzReader;

  /** the current state of the DFA */
  private int zzState;

  /** the current lexical state */
  private int zzLexicalState = YYINITIAL;

  /** this buffer contains the current text to be matched and is
      the source of the yytext() string */
  private char zzBuffer[] = new char[ZZ_BUFFERSIZE];

  /** the textposition at the last accepting state */
  private int zzMarkedPos;

  /** the current text position in the buffer */
  private int zzCurrentPos;

  /** startRead marks the beginning of the yytext() string in the buffer */
  private int zzStartRead;

  /** endRead marks the last character in the buffer, that has been read
      from input */
  private int zzEndRead;

  /** number of newlines encountered up to the start of the matched text */
  private int yyline;

  /** the number of characters up to the start of the matched text */
  private int yychar;

  /**
   * the number of characters from the last newline up to the start of the 
   * matched text
   */
  private int yycolumn;

  /** 
   * zzAtBOL == true <=> the scanner is currently at the beginning of a line
   */
  private boolean zzAtBOL = true;

  /** zzAtEOF == true <=> the scanner is at the EOF */
  private boolean zzAtEOF;

  /** denotes if the user-EOF-code has already been executed */
  private boolean zzEOFDone;
  
  /** 
   * The number of occupied positions in zzBuffer beyond zzEndRead.
   * When a lead/high surrogate has been read from the input stream
   * into the final zzBuffer position, this will have a value of 1;
   * otherwise, it will have a value of 0.
   */
  private int zzFinalHighSurrogate = 0;

  /* user code: */
  StringBuffer string = new StringBuffer();

  private Parser yyparser;

  public Yylex(String str, Parser yyparser) {
    this(new java.io.StringReader(str));
    this.yyparser = yyparser;
  }


  /**
   * Creates a new scanner
   *
   * @param   in  the java.io.Reader to read input from.
   */
  Yylex(java.io.Reader in) {
    this.zzReader = in;
  }


  /** 
   * Unpacks the compressed character translation table.
   *
   * @param packed   the packed character translation table
   * @return         the unpacked character translation table
   */
  private static char [] zzUnpackCMap(String packed) {
    char [] map = new char[0x110000];
    int i = 0;  /* index in packed string  */
    int j = 0;  /* index in unpacked array */
    while (i < 170) {
      int  count = packed.charAt(i++);
      char value = packed.charAt(i++);
      do map[j++] = value; while (--count > 0);
    }
    return map;
  }


  /**
   * Refills the input buffer.
   *
   * @return      <code>false</code>, iff there was new input.
   * 
   * @exception   java.io.IOException  if any I/O-Error occurs
   */
  private boolean zzRefill() throws java.io.IOException {

    /* first: make room (if you can) */
    if (zzStartRead > 0) {
      zzEndRead += zzFinalHighSurrogate;
      zzFinalHighSurrogate = 0;
      System.arraycopy(zzBuffer, zzStartRead,
                       zzBuffer, 0,
                       zzEndRead-zzStartRead);

      /* translate stored positions */
      zzEndRead-= zzStartRead;
      zzCurrentPos-= zzStartRead;
      zzMarkedPos-= zzStartRead;
      zzStartRead = 0;
    }

    /* is the buffer big enough? */
    if (zzCurrentPos >= zzBuffer.length - zzFinalHighSurrogate) {
      /* if not: blow it up */
      char newBuffer[] = new char[zzBuffer.length*2];
      System.arraycopy(zzBuffer, 0, newBuffer, 0, zzBuffer.length);
      zzBuffer = newBuffer;
      zzEndRead += zzFinalHighSurrogate;
      zzFinalHighSurrogate = 0;
    }

    /* fill the buffer with new input */
    int requested = zzBuffer.length - zzEndRead;
    int numRead = zzReader.read(zzBuffer, zzEndRead, requested);

    /* not supposed to occur according to specification of java.io.Reader */
    if (numRead == 0) {
      throw new java.io.IOException("Reader returned 0 characters. See JFlex examples for workaround.");
    }
    if (numRead > 0) {
      zzEndRead += numRead;
      /* If numRead == requested, we might have requested to few chars to
         encode a full Unicode character. We assume that a Reader would
         otherwise never return half characters. */
      if (numRead == requested) {
        if (Character.isHighSurrogate(zzBuffer[zzEndRead - 1])) {
          --zzEndRead;
          zzFinalHighSurrogate = 1;
        }
      }
      /* potentially more input available */
      return false;
    }

    /* numRead < 0 ==> end of stream */
    return true;
  }

    
  /**
   * Closes the input stream.
   */
  public final void yyclose() throws java.io.IOException {
    zzAtEOF = true;            /* indicate end of file */
    zzEndRead = zzStartRead;  /* invalidate buffer    */

    if (zzReader != null)
      zzReader.close();
  }


  /**
   * Resets the scanner to read from a new input stream.
   * Does not close the old reader.
   *
   * All internal variables are reset, the old input stream 
   * <b>cannot</b> be reused (internal buffer is discarded and lost).
   * Lexical state is set to <tt>ZZ_INITIAL</tt>.
   *
   * Internal scan buffer is resized down to its initial length, if it has grown.
   *
   * @param reader   the new input stream 
   */
  public final void yyreset(java.io.Reader reader) {
    zzReader = reader;
    zzAtBOL  = true;
    zzAtEOF  = false;
    zzEOFDone = false;
    zzEndRead = zzStartRead = 0;
    zzCurrentPos = zzMarkedPos = 0;
    zzFinalHighSurrogate = 0;
    yyline = yychar = yycolumn = 0;
    zzLexicalState = YYINITIAL;
    if (zzBuffer.length > ZZ_BUFFERSIZE)
      zzBuffer = new char[ZZ_BUFFERSIZE];
  }


  /**
   * Returns the current lexical state.
   */
  public final int yystate() {
    return zzLexicalState;
  }


  /**
   * Enters a new lexical state
   *
   * @param newState the new lexical state
   */
  public final void yybegin(int newState) {
    zzLexicalState = newState;
  }


  /**
   * Returns the text matched by the current regular expression.
   */
  public final String yytext() {
    return new String( zzBuffer, zzStartRead, zzMarkedPos-zzStartRead );
  }


  /**
   * Returns the character at position <tt>pos</tt> from the 
   * matched text. 
   * 
   * It is equivalent to yytext().charAt(pos), but faster
   *
   * @param pos the position of the character to fetch. 
   *            A value from 0 to yylength()-1.
   *
   * @return the character at position pos
   */
  public final char yycharat(int pos) {
    return zzBuffer[zzStartRead+pos];
  }


  /**
   * Returns the length of the matched text region.
   */
  public final int yylength() {
    return zzMarkedPos-zzStartRead;
  }


  /**
   * Reports an error that occured while scanning.
   *
   * In a wellformed scanner (no or only correct usage of 
   * yypushback(int) and a match-all fallback rule) this method 
   * will only be called with things that "Can't Possibly Happen".
   * If this method is called, something is seriously wrong
   * (e.g. a JFlex bug producing a faulty scanner etc.).
   *
   * Usual syntax/scanner level error handling should be done
   * in error fallback rules.
   *
   * @param   errorCode  the code of the errormessage to display
   */
  private void zzScanError(int errorCode) {
    String message;
    try {
      message = ZZ_ERROR_MSG[errorCode];
    }
    catch (ArrayIndexOutOfBoundsException e) {
      message = ZZ_ERROR_MSG[ZZ_UNKNOWN_ERROR];
    }

    throw new Error(message);
  } 


  /**
   * Pushes the specified amount of characters back into the input stream.
   *
   * They will be read again by then next call of the scanning method
   *
   * @param number  the number of characters to be read again.
   *                This number must not be greater than yylength()!
   */
  public void yypushback(int number)  {
    if ( number > yylength() )
      zzScanError(ZZ_PUSHBACK_2BIG);

    zzMarkedPos -= number;
  }


  /**
   * Contains user EOF-code, which will be executed exactly once,
   * when the end of file is reached
   */
  private void zzDoEOF() throws java.io.IOException {
    if (!zzEOFDone) {
      zzEOFDone = true;
      yyclose();
    }
  }


  /**
   * Resumes scanning until the next regular expression is matched,
   * the end of input is encountered or an I/O-Error occurs.
   *
   * @return      the next token
   * @exception   java.io.IOException  if any I/O-Error occurs
   */
  public int yylex() throws java.io.IOException {
    int zzInput;
    int zzAction;

    // cached fields:
    int zzCurrentPosL;
    int zzMarkedPosL;
    int zzEndReadL = zzEndRead;
    char [] zzBufferL = zzBuffer;
    char [] zzCMapL = ZZ_CMAP;

    int [] zzTransL = ZZ_TRANS;
    int [] zzRowMapL = ZZ_ROWMAP;
    int [] zzAttrL = ZZ_ATTRIBUTE;

    while (true) {
      zzMarkedPosL = zzMarkedPos;

      zzAction = -1;

      zzCurrentPosL = zzCurrentPos = zzStartRead = zzMarkedPosL;
  
      zzState = ZZ_LEXSTATE[zzLexicalState];

      // set up zzAction for empty match case:
      int zzAttributes = zzAttrL[zzState];
      if ( (zzAttributes & 1) == 1 ) {
        zzAction = zzState;
      }


      zzForAction: {
        while (true) {
    
          if (zzCurrentPosL < zzEndReadL) {
            zzInput = Character.codePointAt(zzBufferL, zzCurrentPosL, zzEndReadL);
            zzCurrentPosL += Character.charCount(zzInput);
          }
          else if (zzAtEOF) {
            zzInput = YYEOF;
            break zzForAction;
          }
          else {
            // store back cached positions
            zzCurrentPos  = zzCurrentPosL;
            zzMarkedPos   = zzMarkedPosL;
            boolean eof = zzRefill();
            // get translated positions and possibly new buffer
            zzCurrentPosL  = zzCurrentPos;
            zzMarkedPosL   = zzMarkedPos;
            zzBufferL      = zzBuffer;
            zzEndReadL     = zzEndRead;
            if (eof) {
              zzInput = YYEOF;
              break zzForAction;
            }
            else {
              zzInput = Character.codePointAt(zzBufferL, zzCurrentPosL, zzEndReadL);
              zzCurrentPosL += Character.charCount(zzInput);
            }
          }
          int zzNext = zzTransL[ zzRowMapL[zzState] + zzCMapL[zzInput] ];
          if (zzNext == -1) break zzForAction;
          zzState = zzNext;

          zzAttributes = zzAttrL[zzState];
          if ( (zzAttributes & 1) == 1 ) {
            zzAction = zzState;
            zzMarkedPosL = zzCurrentPosL;
            if ( (zzAttributes & 8) == 8 ) break zzForAction;
          }

        }
      }

      // store back cached position
      zzMarkedPos = zzMarkedPosL;

      if (zzInput == YYEOF && zzStartRead == zzCurrentPos) {
        zzAtEOF = true;
            zzDoEOF();
          { return 0; }
      }
      else {
        switch (zzAction < 0 ? zzAction : ZZ_ACTION[zzAction]) {
          case 1: 
            { yyparser.yylval = new ParserVal(new IdentifierLiteral(yytext())); return Parser.IDENTIFIER;
            }
          case 36: break;
          case 2: 
            { yyparser.yylval = new ParserVal(Double.parseDouble(yytext())); return Parser.NUMBER;
            }
          case 37: break;
          case 3: 
            { yybegin(IDENTIFIER); string.setLength(0);
            }
          case 38: break;
          case 4: 
            { 
            }
          case 39: break;
          case 5: 
            { System.err.println("Error: unexpected character '"+yytext()+"'"); return -1;
            }
          case 40: break;
          case 6: 
            { yybegin(STRING); string.setLength(0);
            }
          case 41: break;
          case 7: 
            { return (int) yycharat(0);
            }
          case 42: break;
          case 8: 
            { return Parser.EQ;
            }
          case 43: break;
          case 9: 
            { return Parser.LT;
            }
          case 44: break;
          case 10: 
            { return Parser.GT;
            }
          case 45: break;
          case 11: 
            { string.append( yytext() );
            }
          case 46: break;
          case 12: 
            { yybegin(YYINITIAL); yyparser.yylval = new ParserVal(new StringLiteral(string.toString())); return Parser.STRING;
            }
          case 47: break;
          case 13: 
            { yybegin(YYINITIAL); yyparser.yylval = new ParserVal(new IdentifierLiteral(string.toString())); return Parser.IDENTIFIER;
            }
          case 48: break;
          case 14: 
            { return Parser.OR;
            }
          case 49: break;
          case 15: 
            { return Parser.LE;
            }
          case 50: break;
          case 16: 
            { return Parser.NEQ;
            }
          case 51: break;
          case 17: 
            { return Parser.GE;
            }
          case 52: break;
          case 18: 
            { return Parser.IS;
            }
          case 53: break;
          case 19: 
            { throw new RuntimeException("Illegal escape sequence \""+yytext()+"\"");
            }
          case 54: break;
          case 20: 
            { string.append( '\"' );
            }
          case 55: break;
          case 21: 
            { string.append( '\'' );
            }
          case 56: break;
          case 22: 
            { string.append( '\t' );
            }
          case 57: break;
          case 23: 
            { string.append( '\r' );
            }
          case 58: break;
          case 24: 
            { string.append( '\f' );
            }
          case 59: break;
          case 25: 
            { string.append( '\\' );
            }
          case 60: break;
          case 26: 
            { string.append( '\b' );
            }
          case 61: break;
          case 27: 
            { string.append( '\n' );
            }
          case 62: break;
          case 28: 
            { return Parser.AND;
            }
          case 63: break;
          case 29: 
            { return Parser.NOT;
            }
          case 64: break;
          case 30: 
            { return Parser.NULL;
            }
          case 65: break;
          case 31: 
            { yyparser.yylval = new ParserVal(new BooleanLiteral(true)); return Parser.BOOLEAN;
            }
          case 66: break;
          case 32: 
            { yyparser.yylval = new ParserVal(new BooleanLiteral(false)); return Parser.BOOLEAN;
            }
          case 67: break;
          case 33: 
            { return Parser.INCLUDE;
            }
          case 68: break;
          case 34: 
            { return Parser.END_WITH;
            }
          case 69: break;
          case 35: 
            { return Parser.START_WITH;
            }
          case 70: break;
          default:
            zzScanError(ZZ_NO_MATCH);
        }
      }
    }
  }


}
